library/std/src/sync/mpmc/waker.rs

```diff
@@ -137,66 +149,205 @@ pub(crate) struct SyncWaker {
+        self.inner.lock().unwrap().watch(oper, cx);
+        self.state.park(state.is_waker);
+    }
+
+    /// Unregisters an operation waiting to be ready.
+    #[inline]
+    pub(crate) fn unwatch(&self, oper: Operation) {
+        let mut inner = self.inner.lock().unwrap();
+        inner.unwatch(oper);
+    }
+
     /// Notifies all threads that the channel is disconnected.
     #[inline]
     pub(crate) fn disconnect(&self) {
         let mut inner = self.inner.lock().unwrap();
         inner.disconnect();
-        self.is_empty
-            .store(inner.selectors.is_empty() && inner.observers.is_empty(), Ordering::SeqCst);
     }
 }
 
 impl Drop for SyncWaker {
     #[inline]
     fn drop(&mut self) {
-        debug_assert!(self.is_empty.load(Ordering::SeqCst));
+        debug_assert!(!self.state.has_waiters());
+    }
+}
+
+/// A guard that manages the state of a blocking operation.
+pub(crate) struct BlockingState<'a> {
+    /// True if this thread is the waker thread, meaning it must
+    /// try to notify waiters after it completes.
+    is_waker: bool,
+
+    waker: &'a SyncWaker,
+}
+
+impl BlockingState<'_> {
+    /// Reset the state after waking up from parking.
+    #[inline]
+    pub(crate) fn unpark(&mut self) {
+        self.is_waker = self.waker.state.unpark();
+    }
+}
+
+impl Drop for BlockingState<'_> {
+    fn drop(&mut self) {
+        if self.is_waker && self.waker.state.drop_waker() {
+            self.waker.notify_one();
+        }
+    }
+}
+
+const NOTIFIED: u64 = 0b01;
+const WAKER: u64 = 0b10;
+
+/// The state of a `SyncWaker`.
+///
+/// A waker manages a count of waiting threads, as well as a "waker thread".
+/// The waker thread has the task of waking up another thread after it succeeds,
+/// transferring ownership of notifications. If there is an active waker thread,
+/// additional threads are not notified. This throttles wakeups, reducing contention
+/// on the channel, as well as creates a chain that makes up for lost wakeups,
+/// allowing threads to park even if the channel is in an inconsistent state.
+struct WakerState {
+    state: AtomicU64,
+}
+
+impl WakerState {
+    /// Initialize the waker state.
+    fn new() -> WakerState {
+        WakerState { state: AtomicU64::new(0) }
+    }
+
+    /// Returns whether or not a waiter needs to be notified.
+    ///
+    /// Note that the preceding write to the channel must use `SeqCst` ordering.
+    fn try_notify(&self) -> bool {
+        // Because writes to the channel are also sequentially consistent,
+        // this creates a total order between storing a value and registering a waiter.
+        let state = self.state.load(Ordering::SeqCst);
+
+        // If a notification is already set, the waker thread will take care
+        // of further notifications. Otherwise we have to notify if there are waiters.
+        if (state >> WAKER) > 0 && (state & NOTIFIED == 0) {
+            return self
```

Comment:
given the fetch_update() is likely to be a a compare_exchange_weak loop already, could it go straight into it instead of doing this extra load check before it?